### DISCUSSION TITLE
Poetry install in non-editable mode

### DISCUSSION THREAD

**Discussion by Ogaday** [python-poetry/poetry#7394] [Q&A]
Running `poetry install` installs the current project in editable mode by [default](https://python-poetry.org/docs/basic-usage/#installing-dependencies-only). I can't find any way to configure this. Should there be an option to `poetry install` the current project _not_ in editable mode?

My use case is for containerised applications, not libraries. For instance, I was experimenting with the [`poetry-dynamic-versioning`](https://github.com/mtkennerly/poetry-dynamic-versioning) plugin in [this app](https://github.com/Ogaday/demo-app), and because I couldn't install in non-editable mode, the process felt a bit awkward. For instance, I could `poetry build --format wheel && pip install ./dist/* && rm -r ./dist` but it would be nice just to be able to `poetry install --no-editable` in my docker build. In my case, it's because of the quirks of `poetry-dynamic-versioning`, which overwrites version variables in the current package based on git tags but reverts changes after the build stage which works fine for dist files but not when the package is installed in editable mode.

Another advantage for allowing install of applications in non-editable mode is that multistage builds become easier. For instance, I often use multi-stage docker builds, and install my project & deps in the build stage, and copy the `lib` directory to the serving stage. This is more difficult if the current package isn't bundled and installed in `site-packages` along with the dependencies.

Is my use case niche/is there a better way, or could this be a feature?

**Comment by yrro** [comment]
I think your use case is covered by the [bundle plugin](https://github.com/python-poetry/poetry-plugin-bundle).

See also #1382

**Reply by tribals** [reply]
No, it is not because bundle plugin does not support *private PyPIs* - what a *dumb* thing!

**Reply by tribals** [reply]
Poetry is actually hardly ever suited for *production deployments* - what a funny thing! ^_^

```Dockerfile
RUN set -ex; \
    apt update; \
    apt install --no-install-recommends --no-install-suggests --yes python3-pip python3-venv; \
    pip install --user --break-system-packages poetry==${poetry_version}

ARG prod
ARG pkg=/pkg
ARG app=/app

WORKDIR ${pkg}

ENV PATH="/root/.local/bin:${PATH}"

COPY pyproject.toml poetry.lock ${pkg}/

RUN set -ex; \
    python3 -m venv ${app}

RUN --mount=type=secret,id=poetry_auth,target=/root/.config/pypoetry/auth.toml \
    . ${app}/bin/activate; \
    poetry install --only=main${prod:+,prod} --no-root

COPY src/ ${pkg}/src/

RUN set -ex; \
    poetry build --format=wheel; \
    ${app}/bin/pip install dist/*.whl
```

```yaml
# compose.yaml
secrets:
  poetry_auth:
    file: ~/.config/pypoetry/auth.toml
```

You see? This should not be that ugly... Don't believe me? *Try* to supply better one!

**Comment by nigelsim** [comment]
Based on @yrro comment I found the following works. Not super elegant, but it is functional and reliable.

`appmodule` is the module folder and `appscript` is a script I have in the `tools.poetry.scripts` section.

```dockerfile
FROM python:3.12-slim-bookworm as builder

RUN pip install poetry==1.7.1

ENV POETRY_NO_INTERACTION=1 \
    POETRY_CACHE_DIR=/tmp/poetry_cache

RUN poetry self add poetry-plugin-bundle

WORKDIR /src

COPY pyproject.toml poetry.lock README.md ./
# Copy the module separately so it ends up in the correct directory structure
COPY appmodule ./appmodule

RUN --mount=type=cache,target=$POETRY_CACHE_DIR poetry bundle venv /src/.venv

FROM python:3.12-slim-bookworm as runtime

ENV VIRTUAL_ENV=/src/.venv \
    PATH="/src/.venv/bin:$PATH"

COPY --from=builder ${VIRTUAL_ENV} ${VIRTUAL_ENV}

ENTRYPOINT ["appscript"]
```
